---
layout: default
---

## Overview 
This course will provide a thorough introduction to microbial community data analysis (metagenomics, metatranscriptomics,
and other culture-independent molecular data) through a balanced approach of lectures and hands-on lab sessions. 
Course participants will learn how to process data from raw meta’omic sequencing files through appropriate bioinformatic methods 
and approaches for subsequent integrative statistical analyses. Participants are invited to bring their own data to the practical 
session on the final day or can use publicly available data from the Integrative Human Microbiome Project (HMP2).

## Format
This course is designed for researchers and students with interest in using culture-independent molecular data (particularly DNA and RNA sequencing technologies) to study microbial communities. This includes both the human microbiome in population studies and techniques generalizable to any microbial communities. The course will mainly focus on the analysis of meta’omic sequencing, including workflows for processing raw sequencing data, multivariate analysis of microbial profiles, and visualization techniques.

## Assumed Background
The participants should have some basic background in microbiology and/or bioinformatics. Programming experience is advantageous but not required, and a basic introduction to UNIX-based command line applications and R will be provided. All labs/tutorials will be run using pre-built cloud instances provided to students. Statistical analyses and visualizations will also be run in R using RStudio.